### PR TITLE
catalog_to_dd speedup & parallelization

### PR DESCRIPTION
**Pull request by flixha**
Thank your for contributing to EQcorrscan!

### What does this PR do?
Allows catalog_to_dd function to run master-slave comparison (correlation) in parallel, and  speeds up that function by reading all events & waveforms only once to keep them in memory throughout execution.

### Why was it initiated?  Any relevant Issues?
catalog_to_dd-function became very slow (>24 hours) for even an intermediate-sized dataset (~2000 events * 50 picks). Also, that function previously read events and waveforms in a double for-loop over master/slave-events, effectively reading them up to n^2 times. By keeping everything in memory , memory consumption can be larger now (~10 GB with my problem). I have implemented some trace-cutting so that as little waveform as possible is kept in memory. The function will still be quicker in serial mode, and even when the swap-memory on the hard disk has to be used compared to before when ASCII-eventfiles and waveform files had to be reread.

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `develop` for bug fixes.
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] All tests still pass.
  - The relevant test catalog_to_dd_test.py passes and tests both serial and parallel execution. However, I have some local issues with my python not accepting pytest.config in tutorials_test.py
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGES.md`.
- [x] First time contributors have added your name to `CONTRIBUTORS.md`.

### CODE REVIEW

**Comment by calum-chamberlain** [comment]
Thanks for this @flixha, I will try and review this tomorrow - not sure what is going on with the CI at the moment though...? I will try and work that out.

Sorry for changing the base-branch, I realised the instruction in the PR template didn't match what I wanted.  Not an issue because develop and master are up-to-date at the moment.  I've updated the template now.

**Comment by flixha** [comment]
Cheers! Yeah, I was at least expecting to get the stickler-email sometime soon but nothing came yet. And as you mention, it didn't matter for the master vs. develop-branches right now as they were fortunately both in sync. I guess I wondered why it should've been master, but will pull-request to develop if I change any more in the future.

**Comment by calum-chamberlain** [comment]
I haven't had a chance to review this yet, but to me it looks like I should have updated this properly a while back when nordic IO moved over to obspy. How are you finding it to use this? Does it fit with your workflow? I'm kinda keen to re-write this to be file-format agnostic, but I don't know how it would work at the moment, what do you think? Do you think it is worth it?
Also, would you be able to make a minor change somewhere (like adding your name to the contributors list if you want to?) and push to your branch to see if the ci triggers again now? It seemed to be working on other branches earlier today.

**Comment by flixha** [comment]
As I was using Seisan to manage my dataset, this function was helpful as it is, but you're of course right that it would be better if it worked independent of dataset format, i.e. by accepting a catalog and stream objects. It would be easy to take the reading data part out of the function; however, right now I see one limitation here:
With the changes that I have made, each waveform will be read, checked for an associated pick, and trimmed right away. If one was to provide a list containing all the streams linked e.g. in each Sfile for each event, that list could be very large in memory (e.g. because one may have 3 minutes of data for all stations of the network in the waveform file, even though only few stations have picks).
I don't know what's the best way to give this to other users, maybe have the user read in events (Sfiles) and provide a helper function to read in the waveform and do checking + trimming?

I pushed again with minor changes as you suggested, but the stickler-ci seems to have run into the same problem...

**Comment by calum-chamberlain** [comment]
> I pushed again with minor changes as you suggested, but the stickler-ci seems to have run into the same problem...

When I ran `pycodestyle` locally it flagged a lot (mostly fixed in my commit).  I had a really quick response from the stickler team, who noted that we ran into the rate limit because of the amount of comments stickler would have posted. So I think we are all go again now.

**Comment by calum-chamberlain** [comment]
> I don't know what's the best way to give this to other users, maybe have the user read in events (Sfiles) and provide a helper function to read in the waveform and do checking + trimming?

Mmm, you make a good point, the main reason this was simple using seisan was because of the relationship between s-files and waveforms. Maybe something to do down the line, but it would involve more thought.

**Comment by flixha** [comment]
Thanks for reviewing and figuring out the stickler issues!
> When I ran pycodestyle locally it flagged a lot (mostly fixed in my commit). I had a really quick response from the stickler team, who noted that we ran into the rate limit because of the amount of comments stickler would have posted. So I think we are all go again now.

Well, I'll just need to be aware of making less mistakes and running pycodestyle locally then next time.. Sorry for too many style inconsistencies!

> Mmm, you make a good point, the main reason this was simple using seisan was because of the relationship between s-files and waveforms. Maybe something to do down the line, but it would involve more thought.

Personally, I only really know Seisan and how files and the database are handled there. (Well, even Seisan may handle this with links to a continuous database...). So to change this, I may just have to get a better idea about how other programs do usually access the waveforms.

I'm testing the suggested changes locally and then commit again.

File: eqcorrscan/utils/catalog_to_dd.py [hunk 1]
```diff
@@ -408,222 +421,498 @@ def write_correlations(event_list, wavbase, extract_len, pre_pick, shift_len,
...
-                        corr_list.append(cc * cc)
-                    except:
-                        msg = "Couldn't compute correlation correction"
-                        warnings.warn(msg)
+                    break
+                # Correct sampling rate of the slave-trace if it differs for
+                # the master-event (allows comparison between e.g. BH and HH
+                # if one is not present before/after).
+                # Then correct the picks
+                try:
```
[the remark this answers is not in the document]
Comment (flixha) [reply eqcorrscan/utils/catalog_to_dd.py]: I changed the loop such that it catches when "xcorr_pick_correction"  fails, e.g. when too few data points are available for cross-correlation. That function stops with an unspecified exception, so cannot catch anything specific with "except".

**Comment by flixha** [comment]
I was testing the scripts a bit more and came upon issue:

What do we want the script to output in the following cases:
1. There are multiple picks for the same phase of one or both events being correlated.
a) The differential times for all comparisons (i.e. 2 diff-P picks if one event has 2 P-picks, 4 diff-P picks if both events have 2 P-picks)  (that's what happens right now)
b) The differential time for the highest cross-correlation value
2. There are multiple waveforms for the same station and channel orientation (e.g. SHZ and BHZ)
a) calculate and output differential times for all combinations of waveforms that match (i.e. same station and channel orientation)
b) only compare one waveform per event per pick (that is what's done now)

**Comment by calum-chamberlain** [comment]
Good questions!

> 1. There are multiple picks for the same phase of one or both events being correlated.
a) The differential times for all comparisons (i.e. 2 diff-P picks if one event has 2 P-picks, 4 diff-P picks if both events have 2 P-picks) (that's what happens right now)
b) The differential time for the highest cross-correlation value

The option that's happening now is, to me, logically the simplest thing to do (take what a user gives the program and output all the results), but it's probably not the *correct* thing to do, and doesn't give the best input to hypoDD. If possible, could you implement both and have the default as option *a* for backwards compatibility.  These would need to be documented.

> 2. There are multiple waveforms for the same station and channel orientation (e.g. SHZ and BHZ)
a) calculate and output differential times for all combinations of waveforms that match (i.e. same station and channel orientation)
b) only compare one waveform per event per pick (that is what's done now)

This is an issue that really stems from SEISAN's use of two characters for channel naming. I think that the comparisons **should only be made on the channel that the pick was made on** (open for discussion though, but I should have documented that better).  

I wonder if we could/should implement a new argument that defines a channel mapping, e.g.
```python
write_correlations(args, channel_mapper={}):
    #Do the things, somewhere down the line when looping through picks
    for pick in event:
        try:
            channel = channel_mapper[pick.waveform_id.channel_code]
        except KeyError:
            channel = pick.waveform_id.channel_code
        tr = st.select(station=pick.waveform_id.station, channel=channel)
        # Rest of logic, and cope with possibility of multiple location codes :(
```
What do you think? If we implemented that as above it would map the SEISAN pick channel to something concrete, but it would break backwards compatibility (which I am not *that* bothered by here), or we could set a default mapper that does what it does now (something like `channel = pick.waveform_id.channel_code[0] + "?" + pick.waveform_id.channel_code[-1]`) maybe with an `if / else` around the `try / except`?

**Comment by calum-chamberlain** [comment]
P.S. thanks for all your work on this @flixha! If there is anything you want me to do at any point, feel free to ask! 🎉

**Comment by flixha** [comment]
Cheers for your quick answer!

> The option that's happening now is, to me, logically the simplest thing to do (take what a user gives the program and output all the results), but it's probably not the correct thing to do, and doesn't give the best input to hypoDD. If possible, could you implement both and have the default as option a for backwards compatibility. These would need to be documented.

> What do you think? If we implemented that as above it would map the SEISAN pick channel to something concrete, but it would break backwards compatibility (which I am not that bothered by here), or we could set a default mapper that does what it does now (something like channel = pick.waveform_id.channel_code[0] + "?" + pick.waveform_id.channel_code[-1]) maybe with an if / else around the try / except?
I agree, we should just include an option here. 

You're right, some of this comes from the (old) Seisan channel naming, but in this script it compares channels even looser, i.e. it just tries to match channel_code[-1]. 

For me right now, that is desired, because when I handle datasets that span more than a few years it's quite usual that either seismometer or just the sampling rate fo distributed channels changed (e.g. upgrade from short period to broad band, SH to BH to HH etc.). For most local microearthquake earthquake studies using hypoDD, it should be possible to relatively safely ignore the change of seismometer and sampling frequency, because the relevant frequency band (~2-10 Hz) should have a pretty flat response for any seismometer. That would be different for teleseismic hypoDD, and one may have to pay attention to the higher frequencies if any channel is only sampled at 20 Hz; so here it would probably be best to also include an option to either strictly match the whole channel name (but checking for Seisan 2-letter naming), or matching only the channel orientation.

**Comment by calum-chamberlain** [comment]
Sounds like your final option:
> so here it would probably be best to also include an option to either strictly match the whole channel name (but checking for Seisan 2-letter naming), or matching only the channel orientation.

Is the way to go. 

>For most local microearthquake earthquake studies using hypoDD, it should be possible to relatively safely ignore the change of seismometer and sampling frequency, because the relevant frequency band (~2-10 Hz) should have a pretty flat response for any seismometer.

I would generally be cagey about this - everything that I have done when I have replaced an instrument, event with the same make and model of instrument, has always resulted in degraded correlations. Because you weight based on correlation (correlation squared), this *might* effect the final result? Anyway, if we were going to go down the route of checking this more thoroughly then we would need to have an inventory passed as well, and I don't think we should go down that route at the moment.

I think your idea of an option to strictly check, or just check orientation (as it does now) will be good, along with a note in the docstring of something like the above comment, would be enough to at least make users aware of what they are doing.

👍

**Comment by calum-chamberlain** [comment]
Hey @flixha - I've made some changes to how clustering is done so that it makes use of the parallelism we have in the correlation functions (#266), which I think would be well placed to replace things in here.  I'm planning on updating lag-calc and catalog_to_dd soon using those functions.  What do you think about closing this, or would you like to leave it open and I can rebase after I have made those changes to the lower-level interfaces?

**Comment by flixha** [comment]
Hi Calum - yes, I don't mind closing this - certainly better to use the better-integrated parallel functions. Was hoping to get back to things in here for a while, but really need to get my thesis done ;-)

**Comment by calum-chamberlain** [comment]
Thanks for that @flixha - I think the things in here will still be worthwhile and I will try and pick them up again soon!
Thesis definitely takes priority! Good luck!

**Pull request closed without merging**